### PR TITLE
MEPTS 1347 | TX NEW 2023 December Release Changes

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 1]
```diff
@@ -237,4 +240,487 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+        + "        WHERE  e.encounter_type = ${52} "
+        + "               AND o.concept_id = ${23866} "
+        + "               AND o.value_datetime <= :endDate "
+        + "               AND o.voided = 0 "
+        + "               AND e.location_id = :location "
+        + "               AND e.voided = 0 "
+        + "               AND p.voided = 0 "
+        + "        GROUP  BY p.patient_id) art "
+        + "GROUP  BY art.patient_id";
+  }
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Already implemented in common queries

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 2]
```diff
@@ -237,4 +240,487 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+            + " FROM ( "
+            + getPatientsArtStartOnFilaOrArvPickup()
+            + "       ) start "
+            + " WHERE start.first_pickup >= '2023-12-21' "
+            + " ) new_art";
+
+    StringSubstitutor sb = new StringSubstitutor(map);
+    cd.setQuery(sb.replace(query));
+    return cd;
+  }
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Not aligned with the specification

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 3]
```diff
@@ -237,4 +240,487 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+        + "                FROM patient p "
+        + "             INNER JOIN encounter e ON e.patient_id = p.patient_id "
+        + "             INNER JOIN ( "
+        + "                   SELECT p.patient_id, MIN(e.encounter_datetime) AS pickup_date "
+        + "                   FROM patient p "
+        + "                   INNER JOIN encounter e ON e.patient_id = p.patient_id "
+        + "                 WHERE e.encounter_type = ${18} "
+        + "                    AND e.voided = 0 "
+        + "                    AND p.voided = 0 "
+        + "                    AND e.location_id = :location "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: End date is missing

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 4]
```diff
@@ -237,4 +240,487 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+        + "       FROM patient p "
+        + "       INNER JOIN encounter e ON e.patient_id = p.patient_id "
+        + "       INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+        + "       WHERE e.encounter_type = ${52} "
+        + "           AND o.concept_id = ${23866} "
+        + "           AND o.voided = 0 "
+        + "           AND e.location_id = :location "
+        + "           AND e.voided = 0 "
+        + "           AND p.voided = 0 "
+        + "       GROUP BY p.patient_id "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: End date is missing

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 5]
```diff
@@ -237,4 +240,487 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+            + "FROM   patient p "
+            + "       INNER JOIN encounter e "
+            + "               ON p.patient_id = e.patient_id "
+            + "       INNER JOIN obs o "
+            + "               ON o.encounter_id = e.encounter_id "
+            + "       INNER JOIN ( "
+            + "       SELECT start.patient_id, "
+            + "              start.first_pickup AS first_pickup "
+            + "             FROM ( "
+            + getPatientsArtStartOnFilaOrArvPickup()
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Art start already implemented in common queries

**Review by eliazmuchave — CHANGES_REQUESTED** [review]
Left comments. Take a look

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 6]
```diff
@@ -237,4 +241,381 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+    map.put("53", hivMetadata.getMasterCardEncounterType().getEncounterTypeId());
+    map.put("52", hivMetadata.getMasterCardDrugPickupEncounterType().getEncounterTypeId());
+    map.put("23866", hivMetadata.getArtDatePickupMasterCard().getConceptId());
+    map.put("23865", hivMetadata.getArtPickupConcept().getConceptId());
+    map.put("1065", hivMetadata.getPatientFoundYesConcept().getConceptId());
+    map.put("1190", hivMetadata.getARVStartDateConcept().getConceptId());
+    map.put("1255", hivMetadata.getARVPlanConcept().getConceptId());
+    map.put("18", hivMetadata.getARVPharmaciaEncounterType().getEncounterTypeId());
+    map.put("1256", hivMetadata.getStartDrugs().getConceptId());
+    map.put("2", hivMetadata.getARTProgram().getProgramId());
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: 1. No need to map the query. It was already mapped on common query class
2. You never check if the Art Start Date is after report start date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 7]
```diff
@@ -237,4 +241,381 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+        "       SELECT patient_id "
+            + " FROM ( "
+            + commonQueries.getARTStartDate(true)
+            + "       ) start "
+            + " WHERE start.first_pickup >= '2023-12-21' ";
+
+    StringSubstitutor sb = new StringSubstitutor(map);
+    cd.setQuery(sb.replace(query));
+    return cd;
+  }
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: You didn't check if the Art Start Date is after report start date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 8]
```diff
@@ -237,4 +241,381 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+    map.put("53", hivMetadata.getMasterCardEncounterType().getEncounterTypeId());
+    map.put("52", hivMetadata.getMasterCardDrugPickupEncounterType().getEncounterTypeId());
+    map.put("23866", hivMetadata.getArtDatePickupMasterCard().getConceptId());
+    map.put("23865", hivMetadata.getArtPickupConcept().getConceptId());
+    map.put("1065", hivMetadata.getPatientFoundYesConcept().getConceptId());
+    map.put("1190", hivMetadata.getARVStartDateConcept().getConceptId());
+    map.put("1255", hivMetadata.getARVPlanConcept().getConceptId());
+    map.put("18", hivMetadata.getARVPharmaciaEncounterType().getEncounterTypeId());
+    map.put("1256", hivMetadata.getStartDrugs().getConceptId());
+    map.put("2", hivMetadata.getARTProgram().getProgramId());
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: No need of mappings
You have not  checked that the art start date is within report period

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java [hunk 9]
```diff
@@ -889,4 +889,79 @@ public String InitialArtStartDateOverallQuery(String endDate, Integer location)
...
+        + "                                  AND e.location_id = :location "
+        + "                                  AND e.encounter_datetime <= :endDate "
+        + "                           GROUP  BY p.patient_id) first_ever "
+        + "                       ON first_ever.patient_id = p.patient_id "
+        + "        WHERE  p.voided = 0 "
+        + "               AND e.voided = 0 "
+        + "               AND e.encounter_type = ${18} "
+        + "               AND e.location_id = :location "
+        + "               AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+        + "        GROUP  BY p.patient_id "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java]: This query is too complicated for what is intended

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java [hunk 10]
```diff
@@ -889,4 +889,79 @@ public String InitialArtStartDateOverallQuery(String endDate, Integer location)
...
+        + "                           GROUP  BY p.patient_id) first_ever_pickup "
+        + "                       ON first_ever_pickup.patient_id = p.patient_id "
+        + "        WHERE  e.encounter_type = ${52} "
+        + "               AND o.concept_id = ${23866} "
+        + "               AND o.value_datetime BETWEEN :startDate AND :endDate "
+        + "               AND e.location_id = :location "
+        + "               AND p.voided = 0 "
+        + "               AND e.voided = 0 "
+        + "               AND o.voided = 0 "
+        + "        GROUP  BY p.patient_id) first "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java]: This query is too complicated for what is intended

**Review by eliazmuchave — CHANGES_REQUESTED** [review]
Left comments

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java (hunk 6 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java (hunk 7 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java (hunk 8 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java (hunk 9 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java (hunk 10 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java]: Resolved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 11]
```diff
@@ -237,4 +241,350 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+            + "FROM   patient p "
+            + "       INNER JOIN encounter e "
+            + "               ON p.patient_id = e.patient_id "
+            + "       INNER JOIN obs o "
+            + "               ON o.encounter_id = e.encounter_id "
+            + "       INNER JOIN ( "
+            + "       SELECT start.patient_id, "
+            + "              start.first_pickup AS first_pickup "
+            + "             FROM ( "
+            + getPatientsFirstDrugPickup()
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: The CD4 must fall between art start - 90 days and art start + 28 day not ARV Pickup

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 12]
```diff
@@ -237,4 +241,350 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+
+    CohortDefinition cd4Above200 =
+        getCd4Result(
+            AdvancedDiseaseAndTBCascadeCohortQueries.Cd4CountComparison.GreaterThanOrEqualTo200mm3);
+
+    cd.addSearch("txnew", EptsReportUtils.map(txnew, mapping2));
+
+    cd.addSearch("cd4Above200", EptsReportUtils.map(cd4Above200, mapping1));
+
+    cd.setCompositionString("txnew AND cd4Above200");
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: according to requirement, you should exclude those with CD4<200mm3

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 13]
```diff
@@ -237,4 +241,350 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+
+    CohortDefinition cd4Above200 =
+        getCd4Result(
+            AdvancedDiseaseAndTBCascadeCohortQueries.Cd4CountComparison.GreaterThanOrEqualTo200mm3);
+
+    cd.addSearch("txnew", EptsReportUtils.map(txnew, mapping2));
+    cd.addSearch("cd4Under200", EptsReportUtils.map(cd4Under200, mapping1));
+    cd.addSearch("cd4Above200", EptsReportUtils.map(cd4Above200, mapping1));
+
+    cd.setCompositionString("txnew AND NOT (cd4Under200 OR cd4Above200)");
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: should exclude all patient with age < 5 from exclusions

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 14]
```diff
@@ -226,7 +230,7 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
     txNewComposition.addParameter(new Parameter("onOrBefore", "onOrBefore", Date.class));
     txNewComposition.addParameter(new Parameter("location", "location", Location.class));
 
-    CohortDefinition startedART = genericCohorts.getStartedArtOnPeriod(false, true);
+    CohortDefinition startedART = getPatientsStartedArtOnFilaOrArvPickupDuringThePeriod();
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Rename the method

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java [hunk 15]
```diff
@@ -237,4 +241,350 @@ public CohortDefinition getTxNewCompositionCohort(String cohortName) {
...
+   *   <li>Drug pick-up date registered on (FILA)
+   *   <li>Drug pick-up date registered on (Recepção Levantou ARV) – Master Card
+   * </ul>
+   *
+   * <p>AND excluding patients with an earliest ART start date from pick-up and clinical sources
+   * (NEW_FR4.1) that falls before (<) 21 December 2023.
+   *
+   * @return {@link CohortDefinition}
+   */
+  public CohortDefinition getPatientsStartedArtOnFilaOrArvPickupDuringThePeriod() {
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxNewCohortQueries.java]: Rename method

**Review by eliazmuchave — CHANGES_REQUESTED** [review]
Left comments. Take a look

**Review by eliazmuchave — APPROVED** [review]
Looks good to me

**Pull request merged**